spec/services/record_participant_declaration_service_spec.rb

```diff
@@ -5,85 +5,119 @@
+    let(:npq_course) { create(:npq_course, identifier: "npq-leading-teaching") }
+
+    let(:npq_validation_data) do
+      create(:npq_validation_data,
+             npq_lead_provider: npq_lead_provider,
+             npq_course: npq_course,
+             user: user)
     end
+
+    let(:params) do
+      {
+        user_id: user.id,
+        declaration_date: "2021-06-21T08:46:29Z",
+        declaration_type: "started",
+        course_identifier: "npq-leading-teaching",
+        lead_provider_from_token: cpd_lead_provider,
+      }
+    end
+
+    let!(:npq_participant_profile) { create(:participant_profile, :npq, teacher_profile: teacher_profile, validation_data: npq_validation_data) }
+
     it "creates a participant and profile declaration" do
-      expect { described_class.call(npq_params) }.to change { ParticipantDeclaration.count }.by(1).and change { ProfileDeclaration.count }.by(1)
+      expect { described_class.call(params) }.to change { ParticipantDeclaration.count }.by(1).and change { ProfileDeclaration.count }.by(1)
     end
   end
 
   context "when sending event for an ect course" do
-    let(:ect_profile) { create(:early_career_teacher_profile) }
-    let(:mentor_profile) { create(:mentor_profile) }
-    let(:ect_params) do
-      params.merge({ lead_provider_from_token: cpd_lead_provider })
-    end
-    let(:mentor_params) do
-      ect_params.merge({ user_id: mentor_profile.user.id, course_identifier: "ecf-mentor" })
-    end
-    let(:induction_coordinator_params) do
-      ect_params.merge({ user_id: induction_coordinator_profile.user_id })
+    let(:params) do
+      {
+        user_id: user.id,
+        declaration_date: "2021-06-21T08:46:29Z",
+        declaration_type: "started",
+        course_identifier: "ecf-induction",
+        lead_provider_from_token: cpd_lead_provider,
+      }
     end
+
     let(:delivery_partner) { create(:delivery_partner) }
-    let!(:school_cohort) { create(:school_cohort, school: ect_profile.school, cohort: ect_profile.cohort) }
-    let!(:partnership) do
+
+    let(:school_cohort) { create(:school_cohort, school: profile.school, cohort: profile.cohort) }
+
+    let(:partnership) do
       create(:partnership,
-             school: ect_profile.school,
+             school: profile.school,
              lead_provider: cpd_lead_provider.lead_provider,
-             cohort: ect_profile.cohort,
+             cohort: profile.cohort,
              delivery_partner: delivery_partner)
     end
 
     context "when lead providers don't match" do
+      before do
+        params[:lead_provider_from_token] = another_cpd_lead_provider
+      end
+
       it "raises a ParameterMissing error" do
         expect { described_class.call(params) }.to raise_error(ActionController::ParameterMissing)
       end
     end
 
     context "when valid user is an early_career_teacher" do
+      before do
+        school_cohort
+        partnership
+      end
```

Comment:
Can we use `let!` where we define them to ensure they are run before tests?

Reply:
I can't replace the `let` with `let!`. The test will fail as the behaviour and outcome would be different. The tests are written in a quirky way. Keeping them lazy then evaluating when needed seems to me the best solution. Please let me know if you have a better suggestion